From: Zheng Liu <wenqing.lz@taobao.com>
Subject: Fix jbd2 kernel bug in jbd2_journal_commit_transaction+0x120c
Patch-mainline: In house 
References: 

In our DB system, we met with a kernel panic when using ext4. After
A deep investigation, it shows that it is caused by a specific usage
of MySQL.
The upstream commit id is deeeaf13 and e7b04ac0.
For more information, please refer to 
http://blog.donghao.org/2013/07/17/rhel6%E5%86%8D%E6%AC%A1%E5%8F%91%E7%8E%B0jbd2%E7%9A%84bug/.

Signed-off-by: Zheng Liu <wenqing.lz@taobao.com>
Signed-off-by: Tao Ma <boyu.mt@taobao.com>

---
 fs/jbd2/journal.c |   25 ++++++++++++++++++++-----
 1 file changed, 20 insertions(+), 5 deletions(-)

--- a/fs/jbd2/journal.c
+++ b/fs/jbd2/journal.c
@@ -466,22 +466,37 @@ int __jbd2_log_space_left(journal_t *jou
  */
 int __jbd2_log_start_commit(journal_t *journal, tid_t target)
 {
+	/* Return if the txn has already requested to be commited */
+	if (journal->j_commit_request == target)
+		return 0;
+
 	/*
-	 * Are we already doing a recent enough commit?
+	 * The only transaction we can possibly wait upon is the
+	 * currently running transaction (if it exists).  Otherwise,
+	 * the target tid must be on old one.
 	 */
-	if (!tid_geq(journal->j_commit_request, target)) {
+	if (journal->j_running_transaction &&
+	    journal->j_running_transaction->t_tid == target) {
 		/*
-		 * We want a new commit: OK, mark the request and wakup the
+		 * We want a new commit: OK, mark the request and wakeup the
 		 * commit thread.  We do _not_ do the commit ourselves.
 		 */
 
 		journal->j_commit_request = target;
-		jbd_debug(1, "JBD: requesting commit %d/%d\n",
+		jbd_debug(1, "JBD2: requesting commmit %d/%d\n",
 			  journal->j_commit_request,
 			  journal->j_commit_sequence);
 		wake_up(&journal->j_wait_commit);
 		return 1;
-	}
+	} else if (!tid_geq(journal->j_commit_request, target))
+		/* This should never happen, but it it does, preserve
+		 * the evidence before kjournald goes into a loop and
+		 * increments j_commit_sequence beyond all recognition. */
+		WARN_ONCE(1, "JBD2: bad log_start_commit: %u %u %u %u\n",
+			  journal->j_commit_request,
+			  journal->j_commit_sequence,
+			  target, journal->j_running_transaction ?
+			  journal->j_running_transaction->t_tid : 0);
 	return 0;
 }
 
